### PR TITLE
feature/calculate traj point velocities

### PR DESCRIPTION
**Pull request by fmessmer**
implements https://github.com/mojin-robotics/cob4/issues/1277#issuecomment-552875855

this pr (is supposed to) implement the following behavior:
```
sss.move("torso","front") -> current behavior: motion uses `default_vel` from yaml - can be float of list of float
sss.move("torso","front",speed_factor=0.5) -> uses half the speed of `default_vel` from yaml (s.o.)
sss.move("torso","front",speed_factor=2.0) -> uses twice the speed of `default_vel` from yaml (s.o.), but fails in case one joint vel exceeds the respective limit from urdf
sss.move("torso","front",speed_factor=0.0) -> will fail due to zero vel
sss.move("torso","front",speed_factor=-1.0) -> will fail due to negative vel
sss.move("torso","front",urdf_vel=True) -> uses vel limits from urdf
sss.move("torso","front",speed_factor=0.5,urdf_vel=True) -> uses half the vel limits from urdf
sss.move("torso","front",speed_factor=2.0,urdf_vel=True) -> will fail due to exceeding limits
sss.move("torso","front",default_vel=1.0) -> overwrites default_vel from yaml
sss.move("torso","front",default_vel=[n*1.0]) -> overwrites default_vel from yaml
sss.move("torso","front",default_vel=[n!=m*1.0]) -> will fail due to dimension mismatch
sss.move("torso","front",urdf_vel=True,default_vel=1.0) -> will fail due to invalid arguments
```

@LoyVanBeek can you carefully review (you could do it commit-by-commit), test and merge this into your feature branch?

@floweisshardt and @fmessmer will then test final version of https://github.com/ipa320/cob_command_tools/pull/259 on hw again

### CODE REVIEW

File: cob_script_server/src/simple_script_server/simple_script_server.py [hunk 1]
```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
 		point_nr = 0
 		traj_time = 0
 
-		param_string = self.ns_global_prefix + "/" + component_name + "/default_vel"
-		if not rospy.has_param(param_string):
-			default_vel = 0.1 # rad/s
-			rospy.logwarn("parameter %s does not exist on ROS Parameter Server, using default of %f [rad/sec].",param_string,default_vel)
-		else:
-			default_vel = rospy.get_param(param_string)
+		if default_vel: # passed via argument
```
Comment (LoyVanBeek) [inline cob_script_server/src/simple_script_server/simple_script_server.py:602]: The branch complexity of this method is getting bigger and bigger. Maybe we should clean that up into at least a separate method for just determining the desired speed... In a next PR

File: cob_script_server/src/simple_script_server/simple_script_server.py [hunk 2]
```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
...
-					velocities.append(robot_urdf.joint_map[joint_name].limit.velocity)
-				except KeyError:
-					velocities.append(default_vel)
+			rospy.logdebug("using default_vel from urdf_limits")
+			velocities = limit_vel
 		else:
-			velocities = [default_vel for _ in joint_names]
-		rospy.loginfo("Velocities are: {}".format(velocities))
+			rospy.logdebug("using default_vel from argument or parameter server")
+			velocities = list(default_vel)
```
Comment (LoyVanBeek) [inline cob_script_server/src/simple_script_server/simple_script_server.py:642]: Why convert back to list here?

File: cob_script_server/src/simple_script_server/simple_script_server.py [hunk 3]
```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
...
+			rospy.logwarn("parameter %s does not exist on ROS Parameter Server, using default_acc {} [rad^2/sec].".format(param_string,default_acc))
 		else:
-			default_acc = rospy.get_param(param_string)
+			param_acc = rospy.get_param(param_string)
+			if (type(param_acc) is float) or (type(param_acc) is int):
+				default_acc = numpy.array([param_acc for _ in start_pos])
+			elif (type(param_acc) is list) and (len(param_acc) == len(start_pos)) and all(((type(item) is float) or (type(item) is int)) for item in param_acc):
+				default_acc = param_acc
+			else:
+				default_acc = numpy.array([0.1 for _ in start_pos]) # rad/s
```
Comment (LoyVanBeek) [inline cob_script_server/src/simple_script_server/simple_script_server.py]: This acc was 1.0  before: https://github.com/LoyVanBeek/cob_command_tools/pull/1/files#diff-8a495e4be2147d5bae5b371366d8f2c6L619 so going to 0.1 would change the behavior

File: cob_script_server/src/simple_script_server/simple_script_server.py [hunk 4]
```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
...
+					rospy.logwarn("parameter %s has wrong format (must be float/int or list of float/int), using default_vel {} [rad/sec].".format(param_string,default_vel))
+		rospy.logdebug("default_vel: {}".format(default_vel))
+
+		robot_urdf = URDF.from_parameter_server()
+		limit_vel = []
+		for idx, joint_name in enumerate(joint_names):
+			try:
+				limit_vel.append(robot_urdf.joint_map[joint_name].limit.velocity)
+			except KeyError:
+				limit_vel.append(default_vel[i])
```
Comment (LoyVanBeek) [inline cob_script_server/src/simple_script_server/simple_script_server.py]: `i` or `idx`?

**Review by LoyVanBeek — CHANGES_REQUESTED** [review]

File: cob_script_server/src/simple_script_server/simple_script_server.py [hunk 5]
```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
...
-		else:
-			default_vel = rospy.get_param(param_string)
+		if default_vel: # passed via argument
+			rospy.logdebug("using default_vel from argument")
+			if (type(default_vel) is float) or (type(default_vel) is int):
+				default_vel = numpy.array([default_vel for _ in start_pos])
+			elif (type(default_vel) is list) and (len(default_vel) == len(start_pos)) and all(((type(item) is float) or (type(item) is int)) for item in default_vel):
+				default_vel = default_vel
+			else:
+				rospy.logerr("parameter %s has wrong format (must be float/int or list of float/int) with proper dimensions.")
```
Comment (LoyVanBeek) [inline cob_script_server/src/simple_script_server/simple_script_server.py:609]: Missing the parameters it want to log. I get output like
```[ERROR] [1573651938.499181]: parameter %s has wrong format (must be float/int or list of float/int) with proper dimensions.```

File: cob_script_server/src/simple_script_server/simple_script_server.py (hunk 2 above)
Comment (fmessmer) [reply cob_script_server/src/simple_script_server/simple_script_server.py:642]: this way I did not have to re-write what's coming below...:wink:

File: cob_script_server/src/simple_script_server/simple_script_server.py [hunk 6]
```diff
@@ -595,31 +599,68 @@ def compose_trajectory(self, component_name, parameter_name, speed_factor=1.0, u
...
+			velocities = limit_vel
 		else:
-			velocities = [default_vel for _ in joint_names]
-		rospy.loginfo("Velocities are: {}".format(velocities))
+			rospy.logdebug("using default_vel from argument or parameter server")
+			velocities = list(default_vel)
+
+		# check velocity limits
+		desired_vel = numpy.array(velocities)*speed_factor
+		if (numpy.any(desired_vel > numpy.array(limit_vel)) or numpy.any(desired_vel < numpy.zeros_like(desired_vel))):
```
Comment (LoyVanBeek) [inline cob_script_server/src/simple_script_server/simple_script_server.py]: When testing with 
```python
sss.move("arm","home",speed_factor=0.5)
sss.move("arm","home",speed_factor=0.5)
sss.move("arm","home",speed_factor=2.0)
sss.move("arm","home",speed_factor=0.0)
sss.move("arm","home",speed_factor=-1.0)
```
I get this error:
`ValueError: operands could not be broadcast together with shapes (0,) (7,)`
Which I don't understand yet ATM

Comment (LoyVanBeek) [reply cob_script_server/src/simple_script_server/simple_script_server.py]: NVM, cannot reproduce after pulling in latest code

**Comment by LoyVanBeek** [comment]
Current test results (checked boxes mean passing tests):
- [x] ```sss.move("arm","home")``` -> current behavior: motion uses `default_vel` from yaml - can be float of list of float
- [x] ```sss.move("arm","home",speed_factor=0.5)``` -> uses half the speed of `default_vel` from yaml (s.o.)
- [x] ```sss.move("arm","home",speed_factor=2.0)``` -> uses twice the speed of `default_vel` from yaml (s.o.), but fails in case one joint vel exceeds the respective limit from urdf
- [x] ```sss.move("arm","home",speed_factor=0.0)``` -> will fail due to zero vel: **NOPE**: I get an error because something goes to infinity. 0 is passed to calculate_point time and something there ges to inf
- [x] ```sss.move("arm","home",speed_factor=-1.0)``` -> will fail due to negative vel. **Note:**: Error message does not mention limit on zero, would be nicer
- [x] ```sss.move("arm","home",urdf_vel=True)``` -> uses vel limits from urdf
- [x] ```sss.move("arm","home",speed_factor=0.5,urdf_vel=True)``` -> uses half the vel limits from urdf
- [x] ```sss.move("arm","home",speed_factor=2.0,urdf_vel=True)``` -> will fail due to exceeding limits
- [ ] ```sss.move("arm","home",default_vel=1.0)``` -> overwrites default_vel from yaml; **Nope**: `desired velocities [ 1.  1.  1.  1.  1.  1.  1.] exceed velocity limits [ 0.5    0.5    0.5    0.7    0.13   0.087  0.5  ],...aborting`
- [ ] ```sss.move("arm","home",default_vel=[1.0,  1.0,  1.0,  1.0,  1.0,  1.0,  1.0])``` -> overwrites default_vel from yaml: **Nope**: `desired velocities [ 1.  1.  1.  1.  1.  1.  1.] exceed velocity limits [ 0.5    0.5    0.5    0.7    0.13   0.087  0.5  ],...aborting`
- [x] ```sss.move("arm","home",default_vel=[1.0,  1.0,  1.0,  1.0,  1.0,  1.0,])``` -> will fail due to dimension mismatch
- [x] ```sss.move("arm","home",urdf_vel=True,default_vel=1.0)``` -> will fail due to invalid arguments

**Comment by fmessmer** [comment]
>     * [ ]  `sss.move("arm","home",speed_factor=0.0)` -> will fail due to zero vel: **NOPE**: I get an error because something goes to infinity. 0 is passed to calculate_point time and something there ges to inf

ok, I will fix this then :point_right: https://github.com/LoyVanBeek/cob_command_tools/pull/1/commits/a9c0f3b638af704bd733d61069795777513f6566

>     * [ ]  `sss.move("arm","home",default_vel=1.0)` -> overwrites default_vel from yaml; **Nope**: `desired velocities [ 1.  1.  1.  1.  1.  1.  1.] exceed velocity limits [ 0.5    0.5    0.5    0.7    0.13   0.087  0.5  ],...aborting`
> 
>     * [ ]  `sss.move("arm","home",default_vel=[1.0,  1.0,  1.0,  1.0,  1.0,  1.0,  1.0])` -> overwrites default_vel from yaml: **Nope**: `desired velocities [ 1.  1.  1.  1.  1.  1.  1.] exceed velocity limits [ 0.5    0.5    0.5    0.7    0.13   0.087  0.5  ],...aborting`

that's ok, the velocity limits are just too small...let me rephrase the examples:
  * [x]  `sss.move("arm","home",default_vel=0.087)` -> overwrites default_vel from yaml; **Nope**: `desired velocities [ 1.  1.  1.  1.  1.  1.  1.] exceed velocity limits [ 0.5    0.5    0.5    0.7    0.13   0.087  0.5  ],...aborting`
 * [x]  `sss.move("arm","home",default_vel=[0.4,  0.4,  0.4,  0.6,  0.13,  0.087,  0.5])` -> overwrites default_vel from yaml: **Nope**: `desired velocities [ 1.  1.  1.  1.  1.  1.  1.] exceed velocity limits [ 0.5    0.5    0.5    0.7    0.13   0.087  0.5  ],...aborting`

**Comment by fmessmer** [comment]
I've fixed the zero and negative error case in https://github.com/LoyVanBeek/cob_command_tools/pull/1/commits/a9c0f3b638af704bd733d61069795777513f6566

**Review by LoyVanBeek — APPROVED** [review]

File: cob_script_server/src/simple_script_server/simple_script_server.py (hunk 5 above)
Comment (fmessmer) [reply cob_script_server/src/simple_script_server/simple_script_server.py:609]: ups...can you fix this on your branch?
I'm offline in Germersheim today...
...some more testing would be nice, too...lets try to get this feature ready as much as possible, then test on hw and merge soon...
